src/pp.ml

```diff
@@ -22,6 +22,70 @@ module Ast = struct
     | Newline
     | Text of string
     | Tag of 'a * 'a t
+
+  let compare_both (f : 'a -> 'a -> int) (g : 'b -> 'b -> int) (a, b) (c, d) =
+    let r = f a c in
+    if r <> 0 then
+      r
+    else
+      g b d
+
+  let compare_triple f g h (x, y, z) (a, b, c) =
+    compare_both f (compare_both g h) (x, (y, z)) (a, (b, c))
+
+  let compare compare_tag =
```

Comment:
You can define it directly on `Pp.t`

Reply:
I can't compare `Format` so we can just use this together with `to_ast` for Pp.

Reply:
No. that's unnecessarily slow. Just raise if you get `Format`. We should just get rid of that constructor

Reply:
I'll remove `Format` in another PR. I've made `Pp.compare` which will raise `Invalid_argument` if two `Pp.of_fmt` values are provided.

Reply:
That seems sensible, but it is beyond the scope of what I am willing to do at the moment. We should definitely take another look at that in the future.

Reply:
Ok, could you add a CHANGES entry?